shell.nix

```diff
@@ -3,12 +3,34 @@
 { ... }@args:
 let
   tests = import ./tests.nix args;
-  inherit (tests) pkgs;
+  inherit (tests) privatestorage lint-python;
+  inherit (privatestorage) pkgs mach-nix tahoe-lafs zkapauthorizer;
+
+  python-env = mach-nix.mkPython {
+    inherit (zkapauthorizer.meta.mach-nix) python providers;
+    overridesPre = [
+      (
+        self: super: {
+          inherit tahoe-lafs;
+        }
+      )
+    ];
+    requirements =
+      ''
+      ${builtins.readFile ./requirements/test.in}
+      ${zkapauthorizer.requirements}
+      '';
+  };
 in
 pkgs.mkShell {
+  # Avoid leaving .pyc all over the source tree when manually triggering tests
+  # runs.
+  PYTHONDONTWRITEBYTECODE = "1";
+
   buildInputs = [
-    tests.python
-    tests.lint-python
-    pkgs.niv
+    # Provide the linting tools for interactive usage.
+    lint-python
```

Comment:
I haven't yet internalized how black's opinions change over time and across different Python versions.

One thing to note is that `lint-python` pins Python 3.9 whereas `python-env` respects whatever value is passed in for `python` (currently defaulting to 3.9).

What's better?  Keeping the version of black as stable as possible (in that case, should we be pinning a black version in `tests.nix` too)?  Keeping black's Python runtime as stable as possible?  Making sure black's Python runtime matches our preferred deployment Python runtime?

Reply:
Looking at black's FAQ, there is [this answer](https://black.readthedocs.io/en/stable/faq.html#how-stable-is-black-s-style), which suggests it is fairly stable. I think the only thing that makes a difference based on python version, is the target-version.

I was actually planning on suggesting change that in `pyproject.toml` (where it is currently set to `py27`). I think setting it to just `["py39"]` would be appropriate (though https://github.com/psf/black/issues/751 suggests we should maybe list all we want to support). This will perhaps require a new run of black (which will do things like removing `u""` prefixes).

To answer your question, I don't think pinning the python version matters. Theoretically, we should pin the black version, but I think in practice it isn't necessary.

Reply:
Great.  Thanks for the links, especially.

Since switching the pyproject.toml black config will trigger a bunch of reformatting, I'd like to defer that to a follow-up branch (not long, just until after the Tahoe 1.17.1 PR is merged).

Reply:
(https://github.com/PrivateStorageio/ZKAPAuthorizer/issues/283)